File: spring-data-requery/src/main/java/org/springframework/data/requery/domain/AbstractPersistable.java
package org.springframework.data.requery.domain;

import io.requery.Persistable;
import io.requery.Superclass;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.Objects;

/**
 * Abstract Entity class
 *
 * @author debop@coupang.com
 * @since 18. 6. 4
 */
@Superclass
public abstract class AbstractPersistable<ID> extends AbstractValueObject implements Persistable, Serializable {

    @Nullable
    abstract public ID getId();

    @io.requery.Transient
    public boolean isNew() {
        return getId() == null;
    }

    @Override
    public boolean equals(@Nullable Object other) {
        if (other == null)
            return false;

        if (other instanceof AbstractPersistable) {
            AbstractPersistable that = (AbstractPersistable) other;
            return (isNew() && that.isNew())
                   ? hashCode() == other.hashCode()
                   : Objects.equals(getId(), that.getId());

        }

        return false;
    }

    @Override
    public int hashCode() {
        return (getId() != null) ? getId().hashCode() : System.identityHashCode(this);
    }

    @NotNull
    @Override
    @io.requery.Transient
    protected ToStringBuilder buildStringHelper() {
        return super.buildStringHelper()
            .add("id", getId());
    }

    private static final long serialVersionUID = 5460350519810267858L;
}

File: spring-data-requery/src/main/kotlin/org/springframework/data/requery/domain/AbstractPersistable.kt
package org.springframework.data.requery.domain

import io.requery.Persistable
import io.requery.Superclass
import java.io.Serializable

/**
 * Abstract Entity class for Requery
 *
 * @author debop
 */
@Superclass
abstract class AbstractPersistable<ID>: AbstractValueObject(), Persistable, Serializable {

    abstract val id: ID

    val isNew: Boolean get() = (id == null)

    override fun equals(other: Any?): Boolean = when(other) {
        is AbstractPersistable<*> -> {
            if(isNew && other.isNew) hashCode() == other.hashCode()
            else id == other.id
        }
        else -> false
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: System.identityHashCode(this)
    }

    // NOTE: buildStringHelder 도 메소드이므로 @Transient 를 꼭 지정해줘야 한다.
    @io.requery.Transient
    override fun buildStringHelper(): ToStringBuilder {
        return super.buildStringHelper()
            .add("id", id)
    }
}